driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +170,149 @@ public int getMaxWireVersion() {
+
+                    List<T> nextBatch = commandCursorResult.getResults();
+                    if (commandCursorResult.getServerCursor() == null || !nextBatch.isEmpty()) {
+                        callback.onResult(nextBatch, null);
+                        return;
+                    }
+
+                    if (!resourceManager.operable()) {
+                        callback.onResult(emptyList(), null);
+                        return;
+                    }
+
+                    getMoreLoop(connection, commandCursorResult.getServerCursor(), callback);
+        });
     }
 
-    private void getMore(final AsyncConnection connection, final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
-        connection.commandAsync(namespace.getDatabaseName(), asGetMoreCommandDocument(cursor.getId(), connection.getDescription()),
-                NO_OP_FIELD_NAME_VALIDATOR, ReadPreference.primary(), CommandResultDocumentCodec.create(decoder, "nextBatch"),
-                connectionSource.getOperationContext(), new CommandResultSingleResultCallback(connection, cursor, callback));
-   }
+    private CommandCursorResult<T> toCommandCursorResult(final ServerAddress serverAddress, final String fieldNameContainingBatch,
+            final BsonDocument commandCursorDocument) {
+        CommandCursorResult<T> commandCursorResult = new CommandCursorResult<>(serverAddress, fieldNameContainingBatch,
+                commandCursorDocument);
+        logCommandCursorResult(commandCursorResult);
+        return commandCursorResult;
+    }
 
-    private BsonDocument asGetMoreCommandDocument(final long cursorId, final ConnectionDescription connectionDescription) {
-        BsonDocument document = new BsonDocument("getMore", new BsonInt64(cursorId))
-                .append("collection", new BsonString(namespace.getCollectionName()));
+    @ThreadSafe
+    private static final class ResourceManager extends CursorResourceManager<AsyncConnectionSource, AsyncConnection> {
 
-        int batchSizeForGetMoreCommand = Math.abs(getNumberToReturn(limit, this.batchSize, count.get()));
-        if (batchSizeForGetMoreCommand != 0) {
-            document.append("batchSize", new BsonInt32(batchSizeForGetMoreCommand));
+        ResourceManager(
+                final MongoNamespace namespace,
+                final AsyncConnectionSource connectionSource,
+                @Nullable final AsyncConnection connectionToPin,
+                @Nullable final ServerCursor serverCursor) {
+            super(namespace, connectionSource, connectionToPin, serverCursor);
         }
-        if (maxTimeMS != 0) {
-            document.append("maxTimeMS", new BsonInt64(maxTimeMS));
+
+        /**
+         * Thread-safe.
+         * Executes {@code operation} within the {@link #tryStartOperation()}/{@link #endOperation()} bounds.
+         */
+        <R> AsyncCallbackSupplier<R> execute(final AsyncCallbackSupplier<R> callbackSupplier) {
```

Comment:
This method is named `execute`, and is documented as `Executes {@code operation} ...`, but actually it does not execute anything, it just creates a new `AsyncCallbackSupplier` based on the given `callbackSupplier`, which is actually the `operation`. For it to stay true to its documentation and for the code to be similar to `execute` in `CommandBatchCursor`, the method should guard and actually asynchronously execute the operation it is given.

I expressed the proposed code changes in https://github.com/stIncMale/mongo-java-driver/commit/97ebded4ba552ceb569e66e4c73bdf8e2069d27e (haven't run any tests).

Reply:
Cherry-picked and adapted 👍